_posts/2022-04-20-solving-issues-with-extensions.adoc

```diff
@@ -0,0 +1,247 @@
+The source of the problem is that, to ensure binary compatibility, the https://github.com/hub4j/github-api[Hub4j GitHub API] we use in Quarkus GitHub App introduces bridge methods in the bytecode.
+
+For instance, let's take this method of the `GitHub` class of the GitHub API:
+
+[source,java]
+----
+    @WithBridgeMethods(value = GHUser.class)
+    public GHMyself getMyself() throws IOException {
+        client.requireCredential();
+        return setMyself();
+    }
+----
+
+Historically, it used to return a `GHUser` but, in newer versions, it returns a `GHMyself`,
+which broke the binary compatibility.
+
+To restore it and with the help of the `@WithBridgeMethods` annotation, the GitHub API build will create two methods in the bytecode:
+one returning `GHMyself` and one returning `GHUser`.
+This is very useful if you have compiled your application with an old version of the GitHub API and you just want to use the new version without recompiling your application.
+Typically, in the case of Jenkins, you can drop a new version of the GitHub API without having to recompile all the Jenkins plugins using GitHub API.
+
+At the bytecode level, you end up with something equivalent to the following:
+
+[source,java]
+----
+    public GHMyself getMyself() throws IOException {
+        client.requireCredential();
+        return setMyself();
+    }
+
+    public GHUser getMyself() throws IOException {
+        return getMyself(); <1>
+    }
+----
+<1> `invokevirtual` of `getMyself()` that returns `GHMyself`
+
+And if your existing compiled code was calling `GHUser getMyself()`, it will still work after the change of return type.
+
+All in all, we can relate to the why and it's not that big of a deal as it's fully transparent...
+except when you start using Mockito due to https://github.com/raphw/byte-buddy/issues/1162[a ByteBuddy issue]:
+ByteBuddy can get confused if there are several methods with the same signature but different return types.
+
+[NOTE]
+====
+ByteBuddy is an amazing library and do not see this post as a critic of it.
+This is an extreme corner case that doesn't happen with standard bytecode.
+====
+
+This issue was causing our tests to be unreliable because sometimes ByteBuddy was choosing the wrong method to apply Mockito magic.
+
+== How can we work around this?
+
+In the Quarkus GitHub App case, we don't really care about binary compatibility:
+when upgrading to a new version of the GitHub API,
+users will rebuild their application.
+
+So given these bridge methods are problematic,
+one solution would be to get rid of them.
+
+Obviously, we could fork the GitHub API and avoid generating the bridge methods.
+
+But forking and maintaining a fork forever is definitely not something we should consider if we can avoid it.
+Especially since we want to continue benefiting from all the future improvements of the GitHub API.
+
+So could we somehow keep the library standard but have Quarkus adjust the bytecode when building the application?
```

Comment:
```suggestion
So how could we somehow keep the library standard but have Quarkus adjust the bytecode when building the application?
```

Reply:
I kept this one as is as I answer "yes" to the question just the line below :).